internal/controllers/scheduling/op.go

```diff
@@ -0,0 +1,233 @@
+package scheduling
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"hash/fnv"
+	"reflect"
+	"time"
+
+	apiv1 "github.com/Azure/eno/api/v1"
+	"github.com/go-logr/logr"
+	"github.com/google/uuid"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+type op struct {
+	Synthesizer *apiv1.Synthesizer
+	Composition *apiv1.Composition
+	Reason      opReason
+
+	Dispatched time.Time
+
+	id               uuid.UUID // set when patch is built
+	synthRolloutHash []byte    // memoized
+}
+
+func newOp(synth *apiv1.Synthesizer, comp *apiv1.Composition) *op {
+	syn := comp.Status.CurrentSynthesis
+	o := &op{Synthesizer: synth, Composition: comp}
+
+	switch {
+	case comp.DeletionTimestamp != nil || !comp.InputsExist(synth) || comp.InputsOutOfLockstep(synth) || !controllerutil.ContainsFinalizer(comp, "eno.azure.io/cleanup"):
+		return nil
+
+	case syn == nil:
+		o.Reason = initialSynthesisOp
+		return o
+
+	case comp.ShouldForceResynthesis():
+		o.Reason = forcedResynthesisOp
+		return o
+
+	case comp.Synthesizing():
+		return nil
+
+	case syn.ObservedCompositionGeneration != comp.Generation:
+		o.Reason = compositionModifiedOp
+		return o
+
+	case comp.ShouldIgnoreSideEffects():
+		return nil
+	}
+
+	nonDeferredInputChanges, deferredInputChanges := inputChangeCount(synth, comp.Status.InputRevisions, syn.InputRevisions)
```

Comment:
Any particular reason these are outside the switch? I assume that it's beacuse it relies on syn not being nil?

Reply:
Yeah exactly. And the multiple return values make it awkward in the case statement anyway